### ISSUE TITLE
Help！

### ISSUE THREAD

**Issue by ztjjj9** [comparch-security/spike-cache#2]
Hello！
When using Spike, configuring different cache sizes for the same program results in different cache miss rates, but surprisingly, the runtime remains the same. This doesn’t seem logical. I want to obtain approximate runtime values for the program with different cache sizes. Can your Spike-cache tool provide estimated runtimes for different cache sizes? The runtime should vary with cache size changes, so this inconsistency needs clarification. 
Thank you! Looking forward to your response.

**Comment by wsong83** [comment]
The cache model in Spike does not actually store any data. It does not support coherence protocol or multicore.
Increasing the cache size should have a small impact on execution time, mainly caused by extra calculation of the replacement policy.

Our cache model do support coherence and multicore. The time difference should be significantly longer.

**Comment by ztjjj9** [comment]
> The cache model in Spike does not actually store any data. It does not support coherence protocol or multicore. Increasing the cache size should have a small impact on execution time, mainly caused by extra calculation of the replacement policy.
> 
> Our cache model do support coherence and multicore. The time difference should be significantly longer.

That is great!  Thanks!

**Comment by ztjjj9** [comment]
<img width="537" alt="微信图片_20240708164032" src="https://github.com/comparch-security/spike-cache/assets/134563203/56efa4b4-dedb-459f-9bb6-16ea60b98f8f">
RISCV: /opt/riscv  The RISC-V compiler has been installed in /opt/riscv, but there was an error during make. Is there any other software that still needs to be installed? 
config.log
[config.log](https://github.com/user-attachments/files/16125314/config.log)

**Comment by wsong83** [comment]
This repo is also quite old and we have moved to a new one named spike-felxicas (if you like to try).
Anyway, I have just recompiled the repo my self and found some header missing issues (related to newer GNU GCC)
Please pull the latest master branch and then following the steps below:
~~~
git submodule update --init --recursive
cd cache_model
make
cd ..
mkdir build
cd build
../configure --prefix==$(what ever place you like)
make
make install
~~~

**Comment by ztjjj9** [comment]
> This repo is also quite old and we have moved to a new one named spike-felxicas (if you like to try). Anyway, I have just recompiled the repo my self and found some header missing issues (related to newer GNU GCC) Please pull the latest master branch and then following the steps below:
> 
> ```
> git submodule update --init --recursive
> cd cache_model
> make
> cd ..
> mkdir build
> cd build
> ../configure --prefix==$(what ever place you like)
> make
> make install
> ```
I am willing to try, but after downloading all the subdirectories (running git submodule update --init --recursive) in new repo, I found that there is no cache-model directory in the spike-felxicas directory, so I can’t go into that directory to make it. When I continue to execute the steps you mentioned later, an error occurs.

**Comment by ztjjj9** [comment]
This error:
<img width="608" alt="error" src="https://github.com/comparch-security/spike-cache/assets/134563203/cde664b6-8206-4377-9f9a-5a8b3ad2dfd9">

**Comment by wsong83** [comment]
spike-flexicas replaces cache-model to flexicas. The script is for spike-cache rather than spike-flexicas.

For the error, you just need to make again. Sorry we failed to manage the Makefile.in properly.
libflexicas.a is compiled in parallel with spike rather than before, so comes the error.

**Comment by ztjjj9** [comment]
Thanks  for your guidance! I successfully installed spike-flexicas using the following command! 
`git clone https://github.com/comparch-security/spike-flexicas.git`
`git submodule update --init --recursive`
`cd flexicas`
`make`
`cd ..`
`mkdir build`
`cd build`
`../configure --prefix==$(what ever place you like)`
`make`
`make install`
However, when I tried to run hello world, I encountered an error and couldn’t exit. Could it be that the RISC-V compiler and PK version I installed are not compatible with spike-flexicas? Also, I want to conduct experiments on how different cache sizes affect program execution speed under the RISC-V architecture, but I couldn’t find relevant options in spike-flexicas’s Host Options.

<img width="356" alt="error2" src="https://github.com/user-attachments/assets/3a47a9cc-5c15-441e-a484-5496ed8bc9dc">

**Comment by wsong83** [comment]
You will need to install a relatively new riscv-gnu-toolchain from `https://github.con/riscv-collab/riscv-gnu-toolchain`
What we current use right now is commit ae9efcc Merge pull request #1381 from a4lg/document-prerequisite-when-testing
around Dec 11 2023 by Kito Cheng

As for changing the configuration of cache, you will need to understand the source code `spike-cache.cc` in spike-flexicas/flexicas directory. Especially for the following lines: (https://github.com/comparch-security/FlexiCAS/blob/21fc9796437177752c9cbfac56d1dd44c849aeab/spike-cache.cc#L190)
~~~
void init(int ncore, const char *prefix) {
    NC = ncore;
    core_cycle.resize(NC, 0);
    if(prefix) pfc_log_prefix = std::string(prefix); // not currently used by the simple tracer but other advanced tracer may need print out logs
    auto l1d = cache_gen_l1<L1IW, L1WN, void, MetadataBroadcastBase, ReplaceLRU, MSIPolicy, false, false, void, true>(NC, "l1d");
    core_data = get_l1_core_interface(l1d);
    auto l1i = cache_gen_l1<L1IW, L1WN, void, MetadataBroadcastBase, ReplaceLRU, MSIPolicy, false, true, void, true>(NC, "l1i");
    core_inst = get_l1_core_interface(l1i);
    auto l2 = cache_gen_l2_exc<L2IW, L2WN, void, MetadataBroadcastBase, ReplaceSRRIP, MSIPolicy, false, void, true>(NC, "l2");
    auto l3 = cache_gen_llc_inc<L3IW, L3WN, void, MetadataDirectoryBase, ReplaceSRRIP, MESIPolicy, void, true>(NC, "l3");
    auto dispatcher = new SliceDispatcher<SliceHashNorm<> >("disp", NC);
    auto mem = new SimpleMemoryModel<void,void,true>("mem");
    tracer = new SimpleTracer(true);

    for(int i=0; i<NC; i++) {
      l1i[i]->outer->connect(l2[i]->inner, l2[i]->inner->connect(l1i[i]->outer, true));
      l1d[i]->outer->connect(l2[i]->inner, l2[i]->inner->connect(l1d[i]->outer));
      dispatcher->connect(l3[i]->inner);
      l2[i]->outer->connect(dispatcher, l3[0]->inner->connect(l2[i]->outer));
      if(i>0) for(int j=0; j<NC; j++) l3[i]->inner->connect(l2[j]->outer);
      l3[i]->outer->connect(mem, mem->connect(l3[i]->outer));
      l1i[i]->attach_monitor(tracer);
      l1d[i]->attach_monitor(tracer);
      l2[i]->attach_monitor(tracer);
      l3[i]->attach_monitor(tracer);
    }

    mem->attach_monitor(tracer);

#ifdef ENABLE_FLEXICAS_THREAD
    // set up the cache server
    std::thread t(cache_server);
    t.detach();
#endif
  }
~~~

The variables `l1d, l1i, l2, l3` are vectors for the L1-I, L1-D, L2 (private), and L3 (LLC) caches. You can change the parameters of the generation function.

For more ideas on how to change them, see the example codes located in spike-flexicas/flexicas/regression

**Comment by wsong83** [comment]
Side note, if you would like to see the simulation changes along with the cache size, etc, you may like to comment out the macro `ENABLE_FLEXICAS_THREAD` on line 31 of `spike-cache.cc`:
https://github.com/comparch-security/FlexiCAS/blob/21fc9796437177752c9cbfac56d1dd44c849aeab/spike-cache.cc#L31

When defines, the cache model (flexicas) will run on a separate thread in parallel with spike, which means some of the latency caused by cache model is hidden in execution time. If this macro is comment out, the cache model will run in the same thread with spike. Of course the cache latency is then reflected in the execution time of spike-flexicas.

After any changes in spike-cache.cc, you will need to rebuild and install spike-flexicas.

**Comment by ztjjj9** [comment]
Thank you for your help, I will try it out！
